# 02 — Preprocessing
### Topic Modelling on India News Headlines (2001–2023)

**Goal of this notebook:** turn the stratified sample from notebook 01 into text that's ready
for BERTopic modelling.

**Why preprocessing is different for BERTopic vs classical models (LDA/NMF):**
- Classical bag-of-words models (LDA/NMF) need heavy cleaning — stopwords, punctuation, and
  casing are all pure noise to them, since they only see word co-occurrence counts.
- BERTopic's **embedding step** uses a Transformer that was trained on natural sentences —
  punctuation, casing, and word order all carry meaning to it. Over-cleaning before embedding
  can *reduce* embedding quality.
- BERTopic's **topic representation step** (c-TF-IDF, run *after* clustering) is a bag-of-words
  method — this is where stopword removal and normalization genuinely help.

**So we keep two parallel columns:**
1. `text_for_embedding` — lightly cleaned → fed into the sentence embedding model.
2. `text_for_representation` — heavily cleaned/lemmatized → used only for the c-TF-IDF
   vectorizer inside BERTopic, and for coherence evaluation later.


In [19]:
# Uses the same local session storage set up in notebook 01 — this notebook must be run
# in the same continuous Colab session as notebook 01 (or after re-running notebook 01
# first in this session), since /content/ storage doesn't persist across sessions.
import os
# No Google Drive needed anymore — all storage is local to this Colab session.
# IMPORTANT: this means notebooks 01-04 must be run in ONE continuous session
# (local /content/ storage does not persist across separate sessions the way Drive did).
BASE_DIR = '/content/topic-modelling-capstone'
DATA_PROCESSED = f'{BASE_DIR}/data/processed'
os.makedirs(DATA_PROCESSED, exist_ok=True)

# --- Libraries ---
# contractions: expands "won't" -> "will not" etc, for consistency.
# nltk: stopword list, WordNet lemmatizer.
!pip install -q nltk contractions

import pandas as pd
import numpy as np
import re
import contractions
import nltk
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

# Download the NLTK data files these tools rely on (quiet=True suppresses the download log).
nltk.download('stopwords', quiet=True)
nltk.download('wordnet', quiet=True)
nltk.download('omw-1.4', quiet=True)

RANDOM_STATE = 42
np.random.seed(RANDOM_STATE)


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 345.1/345.1 kB 22.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.8/114.8 kB 10.9 MB/s eta 0:00:00


## 1. Load the sampled dataset from notebook 01

In [20]:
# Reads the exact file notebook 01 saved — same local path, so this works as long as
# notebook 01 has been run earlier in this SAME Colab session.
df = pd.read_csv(f'{DATA_PROCESSED}/headlines_sample_300k.csv')
print(f"Loaded {len(df):,} rows")
df.head()


Loaded 300,000 rows


,publish_date,headline_category,headline_text,year,month,year_month,category_top,headline_word_count,headline_char_count
0,2001-01-04,unknown,DU North Zone inter-varsity cricket champions,2001,1,2001-01,unknown,6,45
1,2001-07-09,city.hyderabad,Probe all unanimous poll winners: CPI,2001,7,2001-07,city,6,37
2,2001-09-04,bangalore-times,Vikram; not the Seth; was in town,2001,9,2001-09,bangalore-times,7,33
3,2001-10-30,business.india-business,SAP aims to invest more in India,2001,10,2001-10,business,7,32
4,2001-08-20,entertainment.hindi.bollywood,Anil Kumble: Makes his debut!,2001,8,2001-08,entertainment,5,29


## 2. Deduplicate

News agencies sometimes republish near-identical headlines (wire copy, syndication). Exact
duplicates add no topical information and can bias cluster sizes, so we drop them here.


In [21]:
before = len(df)
# drop_duplicates on headline_text specifically (rather than the whole row) catches headlines
# republished under a different date/category too, not just byte-identical rows.
df = df.drop_duplicates(subset='headline_text').reset_index(drop=True)
print(f"Dropped {before - len(df):,} exact duplicate headlines ({(before-len(df))/before:.2%})")
print(f"Remaining: {len(df):,} rows")


Dropped 5,096 exact duplicate headlines (1.70%)
Remaining: 294,904 rows


## 3. Light cleaning for embeddings (`text_for_embedding`)

Minimal intervention: fix encoding artifacts, expand contractions, normalize whitespace. We
deliberately **keep** casing and punctuation, since the embedding model uses them as signal.


In [22]:
def light_clean(text):
    text = str(text)
    # Occasionally scraped/exported text has leftover HTML entities (&amp; etc) instead of
    # the actual characters — strip these out.
    text = re.sub(r'&amp;|&quot;|&#39;', ' ', text)
    # contractions.fix() expands things like "won't" -> "will not", "it's" -> "it is".
    # Wrapped in try/except because a handful of malformed strings can occasionally trip it up;
    # if that happens we just keep the text as-is rather than losing the row.
    try:
        text = contractions.fix(text)
    except Exception:
        pass
    # Collapse any run of whitespace (multiple spaces, tabs) into a single space, and trim
    # leading/trailing whitespace.
    text = re.sub(r'\s+', ' ', text).strip()
    return text

# .apply() runs light_clean() on every row of headline_text and stores the result in a new column.
df['text_for_embedding'] = df['headline_text'].apply(light_clean)

# Spot check a few before/after examples to make sure the cleaning looks sensible.
df[['headline_text', 'text_for_embedding']].sample(5, random_state=RANDOM_STATE)


,headline_text,text_for_embedding
224878,sabarimala opens today amid unprecedented secu...,sabarimala opens today amid unprecedented secu...
63436,Congress holds Kolkata to ransom,Congress holds Kolkata to ransom
174696,Robbers were hesitant; Kim Kardashian tells po...,Robbers were hesitant; Kim Kardashian tells po...
25140,Punjabi heart surgeon finds acclaim in US,Punjabi heart surgeon finds acclaim in US
184128,Now; get cheaper cancer medicine at Amrit outlet,Now; get cheaper cancer medicine at Amrit outlet


## 4. Heavy cleaning for topic representation (`text_for_representation`)

Lowercase, strip punctuation/digits, remove stopwords, lemmatize. This version is **only** used
for c-TF-IDF topic keyword extraction and for computing coherence scores — never fed to the
embedding model.


In [23]:
# Standard English stopword list from NLTK (the, and, of, is, ...).
stop_words = set(stopwords.words('english'))

# A few extra near-stopwords specific to news headlines — these appear constantly but carry
# no topical meaning of their own (e.g. "said" attributes a quote but doesn't describe a topic).
custom_stopwords = {'said', 'says', 'say', 'today', 'yesterday', 'pm', 'am', 'mr', 'mrs',
                     'news', 'report', 'reports', 'reported'}
stop_words = stop_words.union(custom_stopwords)

# Lemmatizer reduces words to their dictionary base form (e.g. "elections" -> "election",
# "running" -> "running"... verb lemmatization needs a POS tag to fully normalize "running" ->
# "run"; we keep it simple here with default noun-based lemmatization, which is standard for
# topic modelling use cases).
lemmatizer = WordNetLemmatizer()

def heavy_clean(text):
    text = str(text).lower()
    # Replace anything that isn't a lowercase letter or whitespace with a space — this strips
    # punctuation, digits, and any stray symbols in one step.
    text = re.sub(r'[^a-z\s]', ' ', text)
    tokens = text.split()
    # Keep a token only if: it's not a stopword, AND it's longer than 2 characters (drops
    # leftover single/double-letter noise from the regex step above, e.g. stray "s" from "'s").
    tokens = [lemmatizer.lemmatize(t) for t in tokens if t not in stop_words and len(t) > 2]
    return ' '.join(tokens)

df['text_for_representation'] = df['text_for_embedding'].apply(heavy_clean)

df[['text_for_embedding', 'text_for_representation']].sample(5, random_state=RANDOM_STATE)


,text_for_embedding,text_for_representation
224878,sabarimala opens today amid unprecedented secu...,sabarimala open amid unprecedented security
63436,Congress holds Kolkata to ransom,congress hold kolkata ransom
174696,Robbers were hesitant; Kim Kardashian tells po...,robber hesitant kim kardashian tell police
25140,Punjabi heart surgeon finds acclaim in US,punjabi heart surgeon find acclaim
184128,Now; get cheaper cancer medicine at Amrit outlet,get cheaper cancer medicine amrit outlet


## 5. Filter out low-signal documents

After cleaning, some headlines may be left with very few tokens (e.g. originally just names or
dates), which contribute noise rather than topical signal. We drop documents with fewer than
3 tokens in the representation column.


In [24]:
# Count tokens remaining after heavy cleaning, per row.
df['token_count'] = df['text_for_representation'].str.split().apply(len)

before = len(df)
# Keep rows where: the lightly-cleaned text isn't empty/whitespace-only, AND there are at
# least 3 meaningful tokens left after heavy cleaning. Both conditions guard against feeding
# near-empty documents into the embedding/clustering steps later, which tend to just add noise.
df = df[
    (df['text_for_embedding'].str.strip().str.len() > 0) &
    (df['token_count'] >= 3)
].reset_index(drop=True)

print(f"Dropped {before - len(df):,} low-signal rows ({(before-len(df))/before:.2%})")
print(f"Final preprocessed dataset: {len(df):,} rows")

df['token_count'].describe()


Dropped 9,988 low-signal rows (3.39%)
Final preprocessed dataset: 284,916 rows


,token_count
count,284916.000000
mean,5.875258
std,1.850083
min,3.000000
25%,5.000000
50%,6.000000
75%,7.000000
max,19.000000


## 6. Final sanity checks

Before saving, confirm the two text columns look reasonable and that we haven't accidentally
introduced empty strings or NaNs.


In [25]:
print("Nulls in key columns:")
print(df[['text_for_embedding', 'text_for_representation']].isna().sum())

print("\nExample before/after, full pipeline:")
for i in df.sample(3, random_state=RANDOM_STATE).index:
    print(f"  original:       {df.loc[i, 'headline_text']}")
    print(f"  for embedding:  {df.loc[i, 'text_for_embedding']}")
    print(f"  for repr.:      {df.loc[i, 'text_for_representation']}")
    print()


Nulls in key columns:
text_for_embedding         0
text_for_representation    0
dtype: int64

Example before/after, full pipeline:
  original:       Gujarat: NGT sees red over shipyard close to Khijadiya sanctuary
  for embedding:  Gujarat: NGT sees red over shipyard close to Khijadiya sanctuary
  for repr.:      gujarat ngt see red shipyard close khijadiya sanctuary

  original:       Chomping watermelon lowers your BP
  for embedding:  Chomping watermelon lowers your BP
  for repr.:      chomping watermelon lower

  original:       Villagers tie raksha dhaga to trees in Kolhan forests
  for embedding:  Villagers tie raksha dhaga to trees in Kolhan forests
  for repr.:      villager tie raksha dhaga tree kolhan forest



## 7. Save the preprocessed dataset

This is what `03_modelling_bertopic.ipynb` will load directly.


In [26]:
# Keep only the columns downstream notebooks actually need — no point carrying every
# intermediate column (like headline_word_count from notebook 01) forward.
cols_to_keep = ['publish_date', 'year', 'month', 'headline_category', 'category_top',
                'headline_text', 'text_for_embedding', 'text_for_representation', 'token_count']
df_out = df[[c for c in cols_to_keep if c in df.columns]]

df_out.to_csv(f'{DATA_PROCESSED}/headlines_preprocessed.csv', index=False)
print(f"Saved {len(df_out):,} preprocessed rows to {DATA_PROCESSED}/headlines_preprocessed.csv")


Saved 284,916 preprocessed rows to /content/topic-modelling-capstone/data/processed/headlines_preprocessed.csv


## 8. Summary of preprocessing decisions

- Started with **300,000** sampled rows from notebook 01
- Removed **5,096** exact duplicate headlines (**1.70%**) → 294,904 rows remaining
- Removed **9,988** low-signal rows (fewer than 3 tokens after heavy cleaning, **3.39%**)
- **Final preprocessed dataset: 284,916 rows** (~95% retention from the sample)
- Token counts after heavy cleaning: mean 5.88, median 6, range 3–19 — down from the ~7.76-word
  average in the raw headlines (notebook 01), consistent with stopword/short-word removal
- No nulls in either `text_for_embedding` or `text_for_representation` — clean output
- Two parallel text representations kept: lightly-cleaned (`text_for_embedding`) for the
  Transformer embedding step, and heavily-cleaned/lemmatized (`text_for_representation`) for
  topic keyword extraction (c-TF-IDF) and coherence evaluation
- Rationale: over-cleaning before embedding degrades the semantic signal a Transformer relies on;
  under-cleaning before c-TF-IDF produces noisy, uninterpretable topic keyword lists. Splitting
  the two avoids that trade-off.
